### DISCUSSION TITLE
Some problem with this sketch

### DISCUSSION THREAD

**Discussion by carminelau** [eModbus/eModbus#346] [eModbus applications]
```
#include "ModbusClientRTU.h"

//ModbusClientRTU RS485();          // for auto half-duplex
ModbusClientRTU RS485(5);  // use rtsPin to toggle DE/RE in half-duplex

void handleData(ModbusMessage msg, uint32_t token) 
{
  Serial.printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", msg.getServerID(), msg.getFunctionCode(), token, msg.size());
  for (auto& byte : msg) {
    Serial.printf("%02X ", byte);
  }
  Serial.println("");
}

void handleError(Error error, uint32_t token) 
{
  // ModbusError wraps the error code and provides a readable error message for it
  ModbusError me(error);
  Serial.printf("Error response: %02X - %s\n", error, (const char *)me);
}

void setup() {
  Serial.begin(9600);
// Set up Serial2 connected to Modbus RTU
  RTUutils::prepareHardwareSerial(Serial1);
  Serial1.begin(9600, SERIAL_8E1, 7, 6);

// Set up ModbusClientRTU client.
// - provide onData and onError handler functions
  RS485.onDataHandler(&handleData);
  RS485.onErrorHandler(&handleError);

// Start ModbusClientRTU background task
  RS485.begin(Serial1);
}

void loop(){
  Error err = RS485.addRequest(0x12345678, 1, READ_HOLD_REGISTER, 0, 1);
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", err, (const char *)e);
  }

  delay(10000);

}
```

My problem is that I always get these messages:
Error response: E5 - Packet length error
Error response: E2 - CRC check error

how i can resolve it ?

**Comment by Miq1** [comment]
This is no issue but a Discussion as well, so I moved it.

The combination of E5/E2 errors is common with faulty wiring of the RS485 bus. Please check again
- if you connected A+ of your RS485 adapter to A+ at the server
- same for B-
- At one end of the bus biasing resistors have to be placed. 4K7 Ohms will do, one between Vcc and A+, another between GND and B-
- a 120 Ohms termination resistor has to be at both ends between A+ and B-. Normally this is already integrated on the RS485 adapters, but you may have to add one at the server side, if that does not have it built in.

**Reply by carminelau** [reply]
perfect thank you. last question how do i send this request ? 

![image](https://github.com/eModbus/eModbus/assets/56676908/949bddcf-4426-4a4c-a9b6-8ffcd4aec2f2)

**Reply by Miq1** [reply]
```
Error err = RS485.addRequest(0x12345678, 1, READ_HOLD_REGISTER, 0, 38);
  if (err!=SUCCESS) {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", err, (const char *)e);
  }
```

**Reply by carminelau** [reply]
okok, but i see only this how response:
![image](https://github.com/eModbus/eModbus/assets/56676908/41492f53-c4dd-456d-b2ab-8931872fe961)

**Reply by Miq1** [reply]
Looks like you are seeing your own request. Either the cabling induces a loop or the server is set to echo the request.

**Reply by carminelau** [reply]
the wiring is correct, because it works with the old library. How can I tell if there is an echo on the server?

**Reply by carminelau** [reply]
I have connected my slave (anemometer) to an esp32 which must send requests to it and wait for a response.

**Reply by Miq1** [reply]
That will only the server documentation tell you. Are you sure there is no crosswise connection between RX and TX on the ESP32? 

It would help if you could provide a verbose log of the running sketch, not only the output. What development environment are you using - the Arduino IDE or Platformio?

**Reply by carminelau** [reply]
1. Are you sure there is no crosswise connection between RX and TX on the ESP32?
- Yes, i test it with old library and all work

2.  What development environment are you using - the Arduino IDE or Platformio? 
- Now, for the prototype use Arduino IDE, but after this prototype i add this sketch to my firmware on PlatformIO

**Reply by Miq1** [reply]
The you should temporarily modify ``Logging.h`` in the library sources right at the start to set the default logging level ``LOG_LEVEL`` to ``LOG_LEVEL_VERBOSE`` (will be ``LOG_LEVEL_ERROR`` normally), compile and run your sketch again.

**Reply by carminelau** [reply]
now this is the log:

> [V] 35| RTUutils.cpp         [ 112] calculateInterval: Calc interval(9600)=3645
> 10:47:52.149 -> [D] 36| ModbusClientRTU.cpp  [  91] doBegin: Client task 1073451892 started. Interval=3645
> 10:47:52.245 -> [V] 158| ModbusMessage.cpp    [ 477] checkData: Check data #3
> 10:47:52.308 -> [D] 158| ModbusClientRTU.cpp  [ 161] addRequestM: request for 01/03
> 10:47:52.372 -> [D] 281| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
> 10:47:52.437 -> [D] 281| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
> 10:47:52.500 -> [D] 281| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
> 10:47:52.565 -> [D] Sent packet: @3FFB94DC/6:
> 10:47:52.596 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 10:47:52.693 -> [D] 536| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
> 10:47:54.695 -> [D] 2660| RTUutils.cpp         [ 437] receive: C/[D] Received packet: @3FFB94DC/1:
> 10:47:54.758 ->   | 0000: E0                                                |.               |
> 10:47:54.855 -> [D] 2783| ModbusClientRTU.cpp  [ 282] handleConnection: Error response (1 bytes) received.
> 10:47:54.953 -> [V] Data: @3FFB94DC/1:
> 10:47:54.984 ->   | 0000: E0                                                |.               |
> 10:47:55.049 -> [D] 2906| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
> 10:47:55.145 -> [V] Response packet: @3FFB94DC/3:
> 10:47:55.176 ->   | 0000: 01 83 E0                                          |...             |
> 10:47:55.241 -> Error response: E0 - Timeout
> 10:48:02.429 -> [V] 10404| ModbusMessage.cpp    [ 477] checkData: Check data #3
> 10:48:02.493 -> [D] 10404| ModbusClientRTU.cpp  [ 161] addRequestM: request for 01/03
> 10:48:02.557 -> [D] 10526| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
> 10:48:02.621 -> [D] 10526| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
> 10:48:02.685 -> [D] 10526| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
> 10:48:02.749 -> [D] Sent packet: @3FFB94DC/6:
> 10:48:02.781 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 10:48:02.877 -> [D] 10781| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
> 10:48:04.909 -> [D] 12905| RTUutils.cpp         [ 437] receive: C/[D] Received packet: @3FFB94DC/1:
> 10:48:05.005 ->   | 0000: E0                                                |.               |
> 10:48:05.101 -> [D] 13028| ModbusClientRTU.cpp  [ 282] handleConnection: Error response (1 bytes) received.
> 10:48:05.198 -> [V] Data: @3FFB94DC/1:
> 10:48:05.198 ->   | 0000: E0                                                |.               |
> 10:48:05.293 -> [D] 13151| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
> 10:48:05.358 -> [V] Response packet: @3FFB94DC/3:
> 10:48:13.065 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 10:48:13.129 -> [D] 21026| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
> 10:48:15.156 -> [D] 23150| RTUutils.cpp         [ 437] receive: C/[D] Received packet: @3FFB94DC/1:
> 10:48:15.253 ->   | 0000: E0                                                |.               |
> 10:48:15.348 -> [D] 23273| ModbusClientRTU.cpp  [ 282] handleConnection: Error response (1 bytes) received.
> 10:48:15.446 -> [V] Data: @3FFB94DC/1:
> 10:48:15.446 ->   | 0000: E0                                                |.               |
> 10:48:15.545 -> [D] 23396| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
> 10:48:15.640 -> [V] Response packet: @3FFB94DC/3:
> 10:48:15.677 ->   | 0000: 01 83 E0                                          |...             |
> 10:48:15.736 -> Error response: E0 - Timeout

**Reply by Miq1** [reply]
All I can see is E0 TIMEOUT errors, no real response, so how did you get the result from the screenshot you posted?

**Reply by carminelau** [reply]
Sorry, I had not commented on a line of code.

Here is the updated verbose log:


> [V] 35| RTUutils.cpp         [ 112] calculateInterval: Calc interval(9600)=3645
> 11:36:05.380 -> [D] 36| ModbusClientRTU.cpp  [  91] doBegin: Client task 1073451892 started. Interval=3645
> 11:36:05.476 -> [V] 158| ModbusMessage.cpp    [ 477] checkData: Check data #3
> 11:36:05.540 -> [D] 158| ModbusClientRTU.cpp  [ 161] addRequestM: request for 01/03
> 11:36:05.604 -> [D] 281| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
> 11:36:05.668 -> [D] 281| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
> 11:36:05.732 -> [D] 281| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
> 11:36:05.795 -> [D] Sent packet: @3FFB94DC/6:
> 11:36:05.827 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 11:36:05.924 -> [D] 536| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
> 11:36:05.988 -> [V] 663| RTUutils.cpp         [ 262] receive: C/3651us without data after 8
> 11:36:06.051 -> [V] 785| RTUutils.cpp         [ 272] receive: C/[V] Raw buffer received: @3FFB9550/8:
> 11:36:06.148 ->   | 0000: 01 03 00 00 00 26 C4 10                           |.....&..        |
> 11:36:06.244 -> [D] 908| RTUutils.cpp         [ 437] receive: C/[D] Received packet: @3FFB94F0/6:
> 11:36:06.308 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 11:36:06.403 -> [D] 1031| ModbusClientRTU.cpp  [ 282] handleConnection: Data response (6 bytes) received.
> 11:36:06.500 -> [V] Data: @3FFB94F0/6:
> 11:36:06.532 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 11:36:06.595 -> [D] 1277| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
> 11:36:06.692 -> [V] Response packet: @3FFB94F0/6:
> 11:36:06.724 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 11:36:06.787 -> Response: serverID=1, FC=3, Token=12345678, length=6:
> 11:36:06.852 -> 01 03 00 00 00 26

**Comment by Miq1** [comment]
Obviously...
```
11:36:06.051 -> [V] 785| RTUutils.cpp [ 272] receive: C/[V] Raw buffer received: @3FFB9550/8:
11:36:06.148 -> | 0000: 01 03 00 00 00 26 C4 10 |.....&.. |
```
...the request came back in - for whatever reason. The time in between the request left the ESP and the response arriving was ~70ms, sufficient for a real communication to the server taking place. So I would suggest looking for the other devices on the bus or that very server, if there is a reason for echoing requests.

**Reply by carminelau** [reply]
okok, sorry but my server is the ESP32 and the slave is the anemometer right ?

**Reply by Miq1** [reply]
No. The ESP is the client, the anemometer is a server. The use of master/slave is deprecated and not used in the standard any more.

**Reply by carminelau** [reply]
ah ok, i didn't know this. How i can resolve my problem ?

**Reply by carminelau** [reply]
could the problem be that I have the DE and the RE split ?

**Reply by Miq1** [reply]
Do you still have that??? In one of the very first replies I advised you to join both DE and RE on a single GPIO pin and give that number to the lib. There is no way to have the lib handle separate DE/RE pins!

(Well, in fact there is, but that will be outside your current knowledge I am afraid - no offense intended!)

**Reply by carminelau** [reply]
Unfortunately I have a custom board with integrated rs485. Tomorrow I'll try to solder them together. If you explain to me how to modify DE and RE separately for manager I will be able to do it.

**Reply by Miq1** [reply]
Here we go: [Documentation about it](https://emodbus.github.io/modbusclient-rtu-api#modbusclientrturtscallback-func-and-modbusclientrturtscallback-func-uint16_t-queuelimit).

Basically you will have to give a function as a ``RTScallback`` instead of a GPIO number to the constructor call. This function will be called whenever the RE/DE pair needs switching, the regular logic level requested will be given as a bool argument to it. In it, you can do whatever is necessary to switch both pins, like
```
void myREDEtoggle(bool level) {
  digitalWrite(myREpin, level ? HIGH : LOW);
  // Your RS485 adapter my require a delayMicroseconds() with some value here
  digitalWrite(myDEpin, level ? LOW : HIGH);
}
```
Both pins need to be defined as ``OUTPUT`` of course. I may have switched the logic levels required in the example, so it can be ``myREpin`` and ``myDEpin`` need to be swapped.

**Reply by carminelau** [reply]
ok perfect, i need to swap the RE and DE. Now the library don't go in Timout, but the response is the same of request:

> 08:58:08.442 -> [V] 20648| ModbusMessage.cpp    [ 477] checkData: Check data #3
> 08:58:08.505 -> [D] 20649| ModbusClientRTU.cpp  [ 161] addRequestM: request for 01/03
> 08:58:08.569 -> [D] 20771| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
> 08:58:08.603 -> [D] 20771| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
> 08:58:08.665 -> [D] 20771| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
> 08:58:08.761 -> [D] Sent packet: @3FFB94DC/6:
> 08:58:08.793 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 08:58:08.889 -> [D] 21026| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
> 08:58:08.953 -> [V] 21153| RTUutils.cpp         [ 262] receive: C/3648us without data after 8
> 08:58:09.017 -> [V] 21153| RTUutils.cpp         [ 272] receive: C/[V] Raw buffer received: @3FFB9550/8:
> 08:58:09.113 ->   | 0000: 01 03 00 00 00 26 C4 10                           |.....&..        |
> 08:58:09.209 -> [D] 21398| RTUutils.cpp         [ 437] receive: C/[D] Received packet: @3FFB94F0/6:
> 08:58:09.305 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 08:58:09.369 -> [D] 21521| ModbusClientRTU.cpp  [ 282] handleConnection: Data response (6 bytes) received.
> 08:58:09.465 -> [V] Data: @3FFB94F0/6:
> 08:58:09.497 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 08:58:09.561 -> [D] 21767| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
> 08:58:09.658 -> [V] Response packet: @3FFB94F0/6:
> 08:58:09.690 ->   | 0000: 01 03 00 00 00 26                                 |.....&          |
> 08:58:09.786 -> Response: serverID=1, FC=3, Token=12345678, length=6:
> 08:58:09.818 -> 01 03 00 00 00 26

**Reply by Miq1** [reply]
There are still few possibilities:
- loopback between the ESP's RX and TX wires
- server is set to echo any request
- another (PC?) on the bus is sending identical requests as well - there must be only one client on the bus
- faulty RS485 adapter

**Reply by carminelau** [reply]
the strange thing is that everything works with the old library.

**Reply by carminelau** [reply]
`Error err = RS485.addRequest(0x12345678, 1, 0x03, 0, 38);
`

In this row the number 0 and 38 means the first and the last register ?

**Comment by Miq1** [comment]
> the strange thing is that everything works with the old library.

Strange indeed. 

> `Error err = RS485.addRequest(0x12345678, 1, 0x03, 0, 38); `
> 
> In this row the number 0 and 38 means the first and the last register ?

🙄 
No. The Modbus standard is requiring a starting register address (0 in your case, which is strange again, since registers are to be counted 1 and up) and a number of registers (38) as parameters for a READ_HOLDING_REGISTER request.

I would suggest trying with address 1 instead - may be the server does not react on wrong requests with an error response but is echoing the request to signal an error state. I have seen a lot in the past years...

**Reply by carminelau** [reply]
nothing, always echo mode.

**Reply by carminelau** [reply]
[Manuale Anemometro.pdf](https://github.com/eModbus/eModbus/files/14885005/Manuale.Anemometro.pdf)

i resend it to read the registers

**Reply by Miq1** [reply]
Register 1 is in fact addressed by address 0x0000 according to the manual.

You could try to first set the server to "Normal mode" as the manual calls it. In Setup, after the ``Serial1.begin()`` call, try to add 
```
Serial1.print(">!\r\n");
```

It is really non-standard, BTW, to use the Modbus connection for another protocol at the same time as they are doing it...
